File: Cuarto/Programación Distribuida/java_concurrencia3/ejercicio5.java
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.LinkedBlockingQueue;

public class ejercicio5 {
    public static class Data {
        private int puntosDentro;
        private int puntosTotales;

        public Data(int puntosDentro, int puntosTotales) {
            this.puntosDentro = puntosDentro;
            this.puntosTotales = puntosTotales;
        }

        public int getPuntosDentro() {
            return puntosDentro;
        }

        public int getPuntosTotales() {
            return puntosTotales;
        }
    }

    static BlockingQueue<Data> queue = new LinkedBlockingQueue<>(10);

    public static class Consumidor extends Thread{
        private int puntosDentro;
        private int puntosTotales;
        public Consumidor() {}

        public void run(){
            while(true){
                try {
                    Data data = queue.take();
                    puntosDentro += data.getPuntosDentro();
                    puntosTotales += data.getPuntosTotales();
                    System.out.println("Puntos dentro: " + puntosDentro + " Puntos totales: " + puntosTotales);
                    System.out.println("Pi: " + (4.0 * puntosDentro / puntosTotales) + " Error: " + (Math.PI - (4.0 * puntosDentro / puntosTotales)));
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }
            }
        }
    }

    public static class Productor extends Thread{
        private int puntos;
        public Productor(int puntos) {
            this.puntos = puntos;
        }

        public void run(){
            while(true){
                double x, y;
                int puntosDentro = 0;
                for (int i = 0; i < puntos; i++){
                    x = Math.random();
                    y = Math.random();
                    if (x*x + y*y <= 1) puntosDentro++;
                }
                try {
                    queue.put(new Data(puntosDentro, puntos));
                    sleep(1000);
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }
            }
        }
    }

    public static void main(String[] args) {
        Consumidor consumidor = new Consumidor();
        for (int i = 0; i < 10; i++) {
            Productor productor = new Productor(100000);
            productor.start();
        }
        consumidor.start();
    }
}

File: Cuarto/ProgramaciónDistribuida/python_concurrencia3/ejercicio5.py
import threading
import math
import random
import time
import queue

class Consumidor (threading.Thread):
    def __init__ (self, cola):
        super().__init__()
        self.cola = cola
        self.puntos_dentro = 0
        self.puntos_totales = 0
    
    def run (self):
        while True:
            data = self.cola.get()
            self.puntos_dentro += data[0]
            self.puntos_totales += data[1]
            print("Puntos dentro del circulo: ", self.puntos_dentro, "Puntos totales: ", self.puntos_totales)
            print("Pi: ", (self.puntos_dentro/self.puntos_totales)*4, 
                  " Error: ", abs(math.pi - (self.puntos_dentro/self.puntos_totales)*4))

class Productor (threading.Thread):
    def __init__ (self, cola, n):
        super().__init__()
        self.cola = cola
        self.n = n
    
    def run (self):
        while True:
            puntos_dentro = 0
            for i in range(0, self.n):
                x = random.uniform(-1, 1)
                y = random.uniform(-1, 1)
                if x**2 + y**2 <= 1: puntos_dentro += 1
            
            self.cola.put((puntos_dentro, self.n))
            time.sleep(1)

cola = queue.Queue()
consumidor = Consumidor(cola)
for _ in range(0,10):
    productor = Productor(cola, 100000)
    productor.start()

consumidor.start()